README.md
## What

Run [Keycloak](https://www.keycloak.org/) on [Cloudron](https://cloudron.io)

## Why

Keycloak can federate users from Cloudron user directory and acts as a OpenID / SAML Provider.

## Build and Install

- Install Cloudron CLI on your machine: `npm install -g cloudron-cli`.
- Install Docker, and make sure you can push to docker hub, or install the docker registry app in your own Cloudron.
- Log in to your Cloudron using cloudron cli: `cloudron login <my.yourdomain.tld>`.
- Build and publish the docker image: `cloudron build`.
- If you're using your own docker registry, name the image properly,
  like `docker.example-cloudron.tld/john_doe/cloudron-keycloak`.
- Log in to Docker Hub and mark the image as public, if necessary.
- Install the app `cloudron install -l <auth.yourdomain.tld>`
- Look at the logs to see if everything is going as planned.

Refer to the [Cloudron Docs](https://docs.cloudron.io/packaging/cli) for more information.

## Third-party Intellectual Properties

All third-party product names, company names, and their logos belong to their respective owners, and may be their
trademarks or registered trademarks.

manifest/CHANGELOG.md
v22.0.1
Update to Keycloac v22.0.1
---
v21.1.1
Update to Keycloak v21.1.1
---
v20.0.3
Bump to Keycloak v20.0.3
---
v20.0.2
Bump to Keycloak v20.0.2
---
v19.0.2
Bump to Keycloak v19.0.2
---
v19.0.1
Bump to Keycloak v19.0.1
---
v18.0.2
Bump to Keycloak v18.0.2
---
v18.0.1
Bump to Keycloak v18.0.1
---
v18.0.0
Bump to Keycloak v18.0.0
---
v17.0.1
Bump to Keycloak v17.0.1
---
v17.0.0
First release of Keycloak on Quarkus.
---
v1.0.0
Initial Version

start.sh
#!/bin/bash
set -eu pipefail

export KC_HOSTNAME="$CLOUDRON_APP_DOMAIN"
export KC_HTTP_ENABLED="true"

pg_cli() {
  PGPASSWORD=$CLOUDRON_POSTGRESQL_PASSWORD psql \
    -h $CLOUDRON_POSTGRESQL_HOST \
    -p $CLOUDRON_POSTGRESQL_PORT \
    -U $CLOUDRON_POSTGRESQL_USERNAME \
    -d $CLOUDRON_POSTGRESQL_DATABASE -c "$1"
}

KC_NEW_REALM=cloudron

# Check for First Run
if [[ ! -f /app/data/conf/keycloak.conf ]]; then
  echo "First run"

  mkdir -p /app/data/themes /app/data/providers /app/data/conf /app/data/tmp
  cp -R /app/code/themes-orig/* /app/data/themes/
  cp -R /app/code/providers-orig/* /app/data/providers/
  cp -R /app/code/conf-orig/* /app/data/conf/
  chown -R 1000:1000 /app/data/

  /app/code/first_run.sh

  touch /app/data/conf/keycloak.conf

  echo "#Basic settings to run Keycloak in production" >/app/data/conf/keycloak.conf
fi

# These values should be re-set to make Keycloak work
echo "Setting ENV variables to /app/data/conf/keycloak.conf"
crudini --set /app/data/conf/keycloak.conf "" db-username "$CLOUDRON_POSTGRESQL_USERNAME"
crudini --set /app/data/conf/keycloak.conf "" db-password "$CLOUDRON_POSTGRESQL_PASSWORD"
crudini --set /app/data/conf/keycloak.conf "" db-url "jdbc:postgresql://$CLOUDRON_POSTGRESQL_HOST/$CLOUDRON_POSTGRESQL_DATABASE"
crudini --set /app/data/conf/keycloak.conf "" http-enabled "true"
crudini --set /app/data/conf/keycloak.conf "" hostname "$CLOUDRON_APP_DOMAIN"
crudini --set /app/data/conf/keycloak.conf "" http-host "0.0.0.0"
crudini --set /app/data/conf/keycloak.conf "" http-port 8080
crudini --set /app/data/conf/keycloak.conf "" proxy "edge"
crudini --set /app/data/conf/keycloak.conf "" proxy-address-forwarding "true"
crudini --set /app/data/conf/keycloak.conf "" hostname-strict "true"
crudini --set /app/data/conf/keycloak.conf "" cache "local"

# Check for SMTP Configuration
pg_cli "DELETE FROM public.realm_smtp_config WHERE realm_id='$KC_NEW_REALM';"
if [[ -z "${CLOUDRON_MAIL_SMTP_SERVER+x}" ]]; then
  echo "SMTP disabled."
else
  echo "SMTP enabled. Setting up SMTP Server values..."

  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_SMTP_SERVER', 'host');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_SMTP_PORT', 'port');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_SMTP_USERNAME', 'user');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_SMTP_PASSWORD', 'password');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_FROM', 'from');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_FROM_DISPLAY_NAME', 'fromDisplayName');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_FROM', 'envelopeFrom');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_FROM', 'replyTo');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '$CLOUDRON_MAIL_FROM_DISPLAY_NAME', 'replyToDisplayName');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '', 'starttls');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', '', 'ssl');"
  pg_cli "INSERT INTO public.realm_smtp_config (realm_id, value, name) VALUES ('$KC_NEW_REALM', 'true', 'auth');"
fi

# Limit memory usage of Java. See https://docs.cloudron.io/packaging/cheat-sheet/#java
if [[ -f /sys/fs/cgroup/memory/memory.memsw.limit_in_bytes ]]; then
  ALLOWED_MBYTES=$(($(cat /sys/fs/cgroup/memory/memory.memsw.limit_in_bytes) / 2 ** 20))
else
  ALLOWED_MBYTES=1024
fi

if [[ -z "${JAVA_OPTS_APPEND+x}" ]]; then
  export JAVA_OPTS_APPEND="-XX:MaxRAM=${ALLOWED_MBYTES}M"
else
  export JAVA_OPTS_APPEND="-XX:MaxRAM=${ALLOWED_MBYTES}M $JAVA_OPTS"
fi

echo "JAVA_OPTS_APPEND: $JAVA_OPTS_APPEND"

chown -R cloudron:cloudron /app/data/

/usr/local/bin/gosu cloudron:cloudron /app/code/bin/kc.sh -cf /app/data/conf/keycloak.conf start --optimized


dev-scripts/docker-run.sh
#!/bin/sh
# Uncomment for Fresh Run
docker rm -f keycloak_custom

# Build with detailed output
DOCKER_BUILDKIT=0 docker build -t keycloak_custom .

# Run in an environment similar to Cloudron.
# Must access using nginx or similar reverse proxy.

BUILDKIT_PROGRESS=plain docker build -t keycloak_custom . &&
docker run --read-only \
  -v "$(pwd)"/.docker/app/data:/app/data:rw \
  -v "$(pwd)"/.docker/tmp:/tmp:rw \
  -v "$(pwd)"/.docker/run:/run:rw \
  -p 8080:8080 \
  --network localnet \
  -e KC_DB_URL_DATABASE=keycloak \
  -e KC_DB_URL_HOST=172.17.0.1 \
  -e KC_DB_URL_PORT=5432 \
  -e KC_DB_USERNAME=postgres \
  -e KC_DB_PASSWORD=postgres \
  -e KC_DB=postgres \
  -e KC_HOSTNAME=keycloak.localhost \
  -e KC_HTTP_HOST=0.0.0.0 \
  -e KC_HTTP_PORT=8080 \
  -e KEYCLOAK_ADMIN=keycloakadmin \
  -e KEYCLOAK_ADMIN_PASSWORD=keycloakadminpassword \
  -e KC_HTTP_ENABLED=true \
  -e KC_STRICT_HTTPS=true \
  keycloak_custom

first_run.sh
#!/bin/bash
set -eu pipefail

pg_cli() {
  PGPASSWORD=$CLOUDRON_POSTGRESQL_PASSWORD psql \
    -h $CLOUDRON_POSTGRESQL_HOST \
    -p $CLOUDRON_POSTGRESQL_PORT \
    -U $CLOUDRON_POSTGRESQL_USERNAME \
    -d $CLOUDRON_POSTGRESQL_DATABASE -c "$1"
}

export KEYCLOAK_HOME=/app/code
export KEYCLOAK_ADMIN=keycloakadmin
export KEYCLOAK_ADMIN_PASSWORD=keycloakadminpassword

echo "Start keycloak with default params"
/usr/local/bin/gosu cloudron:cloudron /app/code/bin/kc.sh start --optimized \
  --db-url "jdbc:postgresql://$CLOUDRON_POSTGRESQL_HOST/$CLOUDRON_POSTGRESQL_DATABASE" \
  --db-username "$CLOUDRON_POSTGRESQL_USERNAME" \
  --db-password "$CLOUDRON_POSTGRESQL_PASSWORD" >/app/data/firstrun.log &

echo "Waiting for the server to start..."

(tail -f -n0 /app/data/firstrun.log &) | grep -q "Listening on:"

echo ">>> Server started. Removing temporary log & waiting for a few seconds..."
rm -f /app/data/firstrun.log
sleep 2

echo "Generate an Admin token using the default admin credentials"
KC_ADMIN_TOKEN_API="http://localhost:8080/realms/master/protocol/openid-connect/token"
KC_ADMIN_API="http://localhost:8080/admin"
KC_ADMIN_TOKEN=$(curl -s -X POST "$KC_ADMIN_TOKEN_API" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -d "username=$KEYCLOAK_ADMIN" \
  -d "password=$KEYCLOAK_ADMIN_PASSWORD" \
  -d 'grant_type=password' \
  -d 'client_id=admin-cli' | jq -r '.access_token')

echo "Create a new Realm because the default master realm is only recommended for managing Keycloak itself."

KC_REALM=cloudron
KC_REALM_API="http://localhost:8080/admin/realms/$KC_REALM"
curl -s -X POST "$KC_ADMIN_API/realms" \
  -H "Content-Type: application/json;charset=utf-8" \
  -H "Authorization: Bearer $KC_ADMIN_TOKEN" \
  -d '{"enabled":true, "id":"'"$KC_REALM"'", "realm":"'"$KC_REALM"'"}'

echo "Realm created"

echo "Setting SMTP Config..."

CLOUDRON_SMTP_SETTINGS='{
  "host": "'"$CLOUDRON_MAIL_SMTP_SERVER"'",
  "port": "'"$CLOUDRON_MAIL_SMTP_PORT"'",
  "from": "'"$CLOUDRON_MAIL_SMTP_USERNAME"'",
  "fromDisplayName": "'"$CLOUDRON_APP_DOMAIN"'",
  "envelopeFrom": "'"$CLOUDRON_MAIL_SMTP_USERNAME"'",
  "replyTo": "'"$CLOUDRON_MAIL_SMTP_USERNAME"'",
  "replyToDisplayName": "'"$CLOUDRON_APP_DOMAIN"'",
  "auth": "true",
  "user": "'"$CLOUDRON_MAIL_SMTP_USERNAME"'",
  "password": "'"$CLOUDRON_MAIL_SMTP_PASSWORD"'",
  "ssl": "",
  "starttls": ""
}'

# Fetch the current realm information.
REALM_JSON=$(curl -s -X GET "$KC_REALM_API" \
  -H "Accept: application/json" \
  -H "Authorization: Bearer $KC_ADMIN_TOKEN" | jq -r .)

echo "Fetched existing Realm Config"

# Replace the Realm's 'smtpServer' with Cloudron SMTP Setting
NEW_REALM_JSON=$(echo "$REALM_JSON" | jq --argjson ARG_SMTP_JSON "$CLOUDRON_SMTP_SETTINGS" '.smtpServer |= $ARG_SMTP_JSON')

curl -s -X PUT "$KC_REALM_API" \
  -H "Accept: application/json" \
  -H "Content-Type: application/json;charset=utf-8" \
  -H "Authorization: Bearer $KC_ADMIN_TOKEN" \
  -d "$NEW_REALM_JSON"

echo "Modified SMTP Configuration successfully"
echo "Done SMTP Config"

echo "Adding LDAP Provider"

LDAP_PROVIDER_NAME='Cloudron LDAP'
LDAP_FILTER='(objectclass=user)'
LDAP_OBJECT_CLASSES='user,inetorgperson,person'
LDAP_USERNAME_ATTRIBUTE='username'
LDAP_RDN_ATTRIBUTE='uid'
LDAP_UUID_ATTRIBUTE='entryUUID'
LDAP_VENDOR='ad'
LDAP_FIRSTNAME_MAPPER='givenName'

# Only create new provider if none exist.
echo "Checking for existing LDAP providers"
COUNT_LDAP_PROVIDERS_RAW=$(pg_cli "SELECT count(id) FROM component WHERE parent_id='$KC_REALM' and provider_id='ldap' and provider_type='org.keycloak.storage.UserStorageProvider';")
COUNT_LDAP_PROVIDERS=$(echo "$COUNT_LDAP_PROVIDERS_RAW" | head -n 3 | tail -n 1 | xargs)

if [[ "$COUNT_LDAP_PROVIDERS" -eq 0 ]]; then
  # NOTE: After adding a LDAP Provider in Keycloak, the first name of the user is mapped to "cn",  so we need to re-map
  # firstName to givenName. Otherwise, a user named John Doe with uerid 'uid-foobar' will be read as "uid-foobar Joe" by Keycloak.

  # Create a provider, and grab the component ID.
  # This request responds with a "location" header, with the newly created component ID. eg:
  # HTTP/1.1 201 Created
  # ...
  # location: $KC_REALM_API/components/component-id
  # ...
  NEW_LDAP_REQUEST=$(curl -s -D - -X POST "$KC_REALM_API/components" \
    -H "Content-Type: application/json;charset=utf-8" \
    -H "Authorization: Bearer $KC_ADMIN_TOKEN" \
    -d '
    {
      "name":"'"$LDAP_PROVIDER_NAME"'",
      "parentId":"'"$KC_REALM"'",
      "providerId":"ldap",
      "providerType":"org.keycloak.storage.UserStorageProvider",
      "config":{
        "vendor":["'"$LDAP_VENDOR"'"],
        "usernameLDAPAttribute":["'"$LDAP_USERNAME_ATTRIBUTE"'"],
        "rdnLDAPAttribute":["'"$LDAP_RDN_ATTRIBUTE"'"],
        "uuidLDAPAttribute":["'"$LDAP_UUID_ATTRIBUTE"'"],
        "userObjectClasses":["'"$LDAP_OBJECT_CLASSES"'"],
        "customUserSearchFilter":["'"$LDAP_FILTER"'"],
        "connectionUrl":["'"$CLOUDRON_LDAP_URL"'"],
        "usersDn":["'"$CLOUDRON_LDAP_USERS_BASE_DN"'"],
        "bindDn":["'"$CLOUDRON_LDAP_BIND_DN"'"],
        "bindCredential":["'"$CLOUDRON_LDAP_BIND_PASSWORD"'"],
        "enabled":["true"],"priority":["0"],"authType":["simple"],"startTls":[],"fullSyncPeriod":["-1"],"changedSyncPeriod":["-1"],"cachePolicy":["DEFAULT"],"evictionDay":[],"evictionHour":[],"evictionMinute":[],"maxLifespan":[],"batchSizeForSync":["1000"],"editMode":["READ_ONLY"],"importEnabled":["false"],"syncRegistrations":["false"],"usePasswordModifyExtendedOp":[],"searchScope":["1"],"validatePasswordPolicy":["false"],"trustEmail":["true"],"useTruststoreSpi":["ldapsOnly"],"connectionPooling":["true"],"connectionPoolingAuthentication":[],"connectionPoolingDebug":[],"connectionPoolingInitSize":[],"connectionPoolingMaxSize":[],"connectionPoolingPrefSize":[],"connectionPoolingProtocol":[],"connectionPoolingTimeout":[],"connectionTimeout":[],"readTimeout":[],"pagination":["true"],"allowKerberosAuthentication":["false"],"serverPrincipal":[],"keyTab":[],"kerberosRealm":[],"debug":["false"],"useKerberosForPasswordAuthentication":["false"]
      }
    }')

  echo "New LDAP Provider Created"

  # Grab the id from the "Location" header, and remove trailing carriage return from the id
  NEW_LDAP_PROVIDER_ID=$(echo "$NEW_LDAP_REQUEST" | grep -Fi 'Location:' | sed "s|Location: $KC_REALM_API/components/||" | sed "s|\r||g")

  # Fetch the "LDAP Mappers" Component for the newly created LDAP Provider.
  # Response contains all mappers in the form [{id:'',name:'',providerId:'',providerType:'',parentId:'',config:{}}]
  LDAP_MAPPERS=$(curl -s \
    -H "Accept: application/json;charset=utf-8" \
    -H "Authorization: Bearer $KC_ADMIN_TOKEN" \
    "$KC_REALM_API/components?type=org.keycloak.storage.ldap.mappers.LDAPStorageMapper&parent=$NEW_LDAP_PROVIDER_ID")

  echo "Fetched existing LDAP Mappers"

  # We are only interested in the mapper with the name "first name".
  FIRSTNAME_MAPPER_ID=$(echo "$LDAP_MAPPERS" | jq -r '.[] | select(.name=="first name") | .id')
  FIRSTNAME_MAPPER_PARENT_ID=$(echo "$LDAP_MAPPERS" | jq -r '.[] | select(.name=="first name") | .parentId')

  # Modify the "first name" LDAP Mapper to map "firstName" -> "givenName".
  curl -X PUT "$KC_REALM_API/components/$FIRSTNAME_MAPPER_ID" \
    -H "Content-Type: application/json;charset=utf-8" \
    -H "Authorization: Bearer $KC_ADMIN_TOKEN" \
    -d '{
        "id":"'"$FIRSTNAME_MAPPER_ID"'",
        "name":"first name",
        "providerId":"user-attribute-ldap-mapper",
        "providerType":"org.keycloak.storage.ldap.mappers.LDAPStorageMapper",
        "parentId":"'"$FIRSTNAME_MAPPER_PARENT_ID"'",
        "config":{
          "ldap.attribute":["'"$LDAP_FIRSTNAME_MAPPER"'"],"is.mandatory.in.ldap":["true"],"always.read.value.from.ldap":["true"],"read.only":["true"],"user.model.attribute":["firstName"],"attribute.default.value":[],"is.binary.attribute":["false"]
        }
    }'
  echo "Modified LDAP Mapper to map firstName -> givenName"
else
  echo "LDAP provider already exists!"
fi

echo "first run setup complete. restarting keycloak as usual."
killall java
